File: doc_model.py
"""Editor document model  <->  canonical Markdown.

This is the round-trip layer under a structured (WYSIWYG) editor: the editor maps
its contenteditable DOM to/from the *document model* defined here, and the model
serializes to the same light Markdown convention that ``manuscript.parse_markdown``
already consumes -> PDF / EPUB / preview unchanged. The engine stays the single
owner of typography; this module only carries structure + emphasis.

Guarantee (validated by ``test_doc_model.py`` over the sample + every real
manuscript, and proven in the serialization spike before this was built):

    ENGINE FIDELITY   parse_markdown(md) == parse_markdown(to_markdown(from_markdown(md)))
    MODEL STABILITY   from_markdown(md)  == from_markdown(to_markdown(from_markdown(md)))

The first parse normalizes (smart quotes, hard-wrap line-joining), so these are
fixed-point properties, not raw-string equality -- editing then re-rendering never
changes the typeset output, and re-opening a saved manuscript never drifts.

Document model
--------------
A document is a list of blocks. Each block is a plain JSON-able dict:

    {"type": "chapter",  "title": str | None}          # "# ..."   title stored raw
    {"type": "part",     "title": str | None}          # "=== ..." title stored raw
    {"type": "subhead",  "runs": [run, ...]}           # "## ..."
    {"type": "para",     "runs": [run, ...]}           # a paragraph
    {"type": "scene"}                                   # scene break
    {"type": "docblock", "block_type": str,            # "" for a plain ~~~ fence
                         "attrs": {key: value, ...},    # ordered as authored
                         "children": [para, ...]}       # each child is a para block

    run = {"text": str, "bold": bool, "italic": bool}  # emphasis is non-nesting

Chapter and part titles are stored raw (no smartening / emphasis) to mirror the
engine; subheads and paragraphs carry smartened text with emphasis runs.

Literal emphasis / marker characters
------------------------------------
The convention has no escape syntax, so a paragraph a writer *types* literally --
"5 * 3", a ``snake_case`` name, a line starting ``#`` or ``~~~`` -- could re-parse
as emphasis or structure. ``to_markdown`` emits a backslash escape ONLY when the
literal text would otherwise re-parse as something other than itself (see
``_escape_plain``), so real-world prose serializes with zero backslashes. Fully
supporting a *deliberately* typed literal ``*word*`` or line-start marker also
needs the matching (backward-compatible) unescape in ``manuscript`` -- see the
roadmap; it is not required for any existing manuscript.
"""

import re

import manuscript


# Bold before italic, mirroring manuscript._inline's substitution order. Emphasis
# is non-nesting in this convention, so a flat run list is faithful.
_BOLD_RE = re.compile(r'\*\*(.+?)\*\*')
_ITALIC_STAR = re.compile(r'(?<!\*)\*(?!\s)(.+?)(?<!\s)\*(?!\*)')
_ITALIC_UNDER = re.compile(r'_(?!\s)(.+?)(?<!\s)_')


# ---------------------------------------------------------------------------
# Escape layer. A backslash escapes a literal emphasis char; escaped chars are
# parked on private-use codepoints so the emphasis regexes never see them, then
# restored to literals. No manuscript contains a backslash, so this is inert on
# existing content.
# ---------------------------------------------------------------------------

_BSL = chr(92)                 # a single backslash
_PARK_STAR = chr(0xE000)
_PARK_UNDER = chr(0xE001)
_PARK_BSL = chr(0xE002)


def _park_escapes(text):
    text = text.replace(_BSL + _BSL, _PARK_BSL)      # \\  -> literal backslash
    text = text.replace(_BSL + '*', _PARK_STAR)      # \*  -> literal *
    text = text.replace(_BSL + '_', _PARK_UNDER)     # \_  -> literal _
    return text


def _restore_escapes(text):
    return (text.replace(_PARK_STAR, '*')
                .replace(_PARK_UNDER, '_')
                .replace(_PARK_BSL, _BSL))


def _escape_all(text):
    """Escape every literal emphasis char (used inside emphasis spans)."""
    return (text.replace(_BSL, _BSL + _BSL)
                .replace('*', _BSL + '*')
                .replace('_', _BSL + '_'))


def _escape_plain(text):
    """Escape a plain run only if leaving it bare would re-parse as emphasis.

    The empirically-correct minimal rule: don't reason char-by-char about the
    engine's guards (fragile, and coupled to exact regex behaviour) -- just ask
    whether the literal text survives a parse unchanged. If it does, emit it bare.
    Quote-independent, so we probe with smartquotes off.
    """
    escaped_bsl = text.replace(_BSL, _BSL + _BSL)
    if _parse_inline(escaped_bsl, smartquotes=False) == [_run(text)]:
        return escaped_bsl
    return escaped_bsl.replace('*', _BSL + '*').replace('_', _BSL + '_')


# ---------------------------------------------------------------------------
# Inline: Markdown emphasis  <->  runs
# ---------------------------------------------------------------------------

def _run(text, bold=False, italic=False):
    return {"text": text, "bold": bold, "italic": italic}


def _parse_inline(text, smartquotes=True):
    """Smarten (optional), then split into emphasis runs."""
    text = _park_escapes(text)
    if smartquotes:
        text = manuscript._smarten(text)

    runs = []
    # Pass 1: carve out bold spans.
    segments = []                                        # (is_bold, str)
    pos = 0
    for m in _BOLD_RE.finditer(text):
        if m.start() > pos:
            segments.append((False, text[pos:m.start()]))
        segments.append((True, m.group(1)))
        pos = m.end()
    if pos < len(text):
        segments.append((False, text[pos:]))

    # Pass 2: carve out italics within each non-bold segment.
    for is_bold, seg in segments:
        if is_bold:
            runs.append(_run(seg, bold=True))
            continue
        for is_italic, chunk in _split_italics(seg):
            if chunk:
                runs.append(_run(chunk, italic=is_italic))

    runs = _coalesce(runs)
    for r in runs:                                       # parked -> literal chars
        r["text"] = _restore_escapes(r["text"])
    return runs


def _split_italics(seg):
    """Yield (is_italic, str) for a non-bold segment: *...* then _..._ ."""
    out = []
    pos = 0
    for m in _ITALIC_STAR.finditer(seg):
        if m.start() > pos:
            out.append((False, seg[pos:m.start()]))
        out.append((True, m.group(1)))
        pos = m.end()
    if pos < len(seg):
        out.append((False, seg[pos:]))

    final = []
    for is_italic, piece in out:
        if is_italic:
            final.append((True, piece))
            continue
        p2 = 0
        for m in _ITALIC_UNDER.finditer(piece):
            if m.start() > p2:
                final.append((False, piece[p2:m.start()]))
            final.append((True, m.group(1)))
            p2 = m.end()
        if p2 < len(piece):
            final.append((False, piece[p2:]))
    return final


def _coalesce(runs):
    """Merge adjacent runs with identical emphasis (keeps the model canonical)."""
    out = []
    for r in runs:
        if out and out[-1]["bold"] == r["bold"] and out[-1]["italic"] == r["italic"]:
            out[-1]["text"] += r["text"]
        else:
            out.append(dict(r))
    return [r for r in out if r["text"]]


def _runs_to_md(runs):
    """Serialize runs back to Markdown emphasis. _italic_ normalizes to *italic*."""
    parts = []
    for r in runs:
        if r["bold"]:
            parts.append(f"**{_escape_all(r['text'])}**")
        elif r["italic"]:
            parts.append(f"*{_escape_all(r['text'])}*")
        else:
            parts.append(_escape_plain(r["text"]))
    return "".join(parts)


def _block_collision(line):
    """True if a paragraph line would be mistaken for a structural block."""
    return bool(
        manuscript.CHAPTER_RE.match(line)
        or manuscript.SUBHEAD_RE.match(line)
        or manuscript.PART_RE.match(line)
        or manuscript.DOCBLOCK_RE.match(line)
        or manuscript.SCENE_BREAK_RE.match(line)
    )


# ---------------------------------------------------------------------------
# Markdown  ->  document model
# ---------------------------------------------------------------------------

def from_markdown(raw, smartquotes=True):
    """Parse canonical Markdown into the editor document model (list of blocks)."""
    lines = raw.replace('\r\n', '\n').replace('\r', '\n').split('\n')
    blocks = []
    para_buf = []

    in_block = False
    block_children = []
    block_para_buf = []
    block_type = ''
    block_attrs = {}

    def flush_para():
        nonlocal para_buf
        if para_buf:
            joined = ' '.join(s.strip() for s in para_buf).strip()
            if joined:
                blocks.append({"type": "para",
                               "runs": _parse_inline(joined, smartquotes)})
            para_buf = []

    def flush_block_para():
        nonlocal block_para_buf
        if block_para_buf:
            joined = ' '.join(s.strip() for s in block_para_buf).strip()
            if joined:
                block_children.append({"type": "para",
                                       "runs": _parse_inline(joined, smartquotes)})
            block_para_buf = []

    def close_block():
        nonlocal in_block, block_children, block_type, block_attrs
        flush_block_para()
        blocks.append({"type": "docblock", "block_type": block_type,
                       "attrs": dict(block_attrs), "children": list(block_children)})
        block_children = []
        block_type = ''
        block_attrs = {}
        in_block = False

    for line in lines:
        # A leading backslash guarding a block marker => literal paragraph text.
        if not in_block and line[:1] == _BSL and _block_collision(line[1:]):
            para_buf.append(line[1:])
            continue

        if not in_block:
            m_part = manuscript.PART_RE.match(line)
            if m_part:
                flush_para()
                blocks.append({"type": "part",
                               "title": m_part.group(1).strip() or None})
                continue

        m_doc = manuscript.DOCBLOCK_RE.match(line)
        if m_doc:
            if in_block:
                close_block()
            else:
                flush_para()
                block_type, block_attrs = manuscript._parse_block_header(m_doc.group(1))
                in_block = True
            continue

        if in_block:
            if line.strip() == '':
                flush_block_para()
            else:
                block_para_buf.append(line)
            continue

        m_ch = manuscript.CHAPTER_RE.match(line)
        m_sub = manuscript.SUBHEAD_RE.match(line)
        if m_ch:
            flush_para()
            blocks.append({"type": "chapter", "title": m_ch.group(1).strip() or None})
            continue
        if manuscript.SCENE_BREAK_RE.match(line):
            flush_para()
            blocks.append({"type": "scene"})
            continue
        if m_sub:
            flush_para()
            blocks.append({"type": "subhead",
                           "runs": _parse_inline(m_sub.group(1).strip(), smartquotes)})
            continue
        if line.strip() == '':
            flush_para()
        else:
            para_buf.append(line)

    if in_block:
        close_block()
    flush_para()
    return blocks


# ---------------------------------------------------------------------------
# Document model  ->  Markdown
# ---------------------------------------------------------------------------

def to_markdown(blocks):
    """Serialize the editor document model back to canonical Markdown."""
    out = []
    for b in blocks:
        t = b["type"]
        if t == "chapter":
            out.append("# " + (b["title"] or ""))
        elif t == "part":
            out.append("=== " + (b["title"] or ""))
        elif t == "subhead":
            out.append("## " + _runs_to_md(b["runs"]))
        elif t == "para":
            line = _runs_to_md(b["runs"])
            if _block_collision(line):                   # guard a would-be marker
                line = _BSL + line
            out.append(line)
        elif t == "scene":
            out.append("* * *")
        elif t == "docblock":
            header = "~~~"
            if b["block_type"]:
                header += " " + b["block_type"]
                for k, v in b["attrs"].items():
                    header += f' {k}="{v}"'
            out.append(header)
            kids = b["children"]
            for i, kid in enumerate(kids):
                out.append(_runs_to_md(kid["runs"]))
                if i < len(kids) - 1:
                    out.append("")                       # blank line between paras
            out.append("~~~")
        else:
            raise ValueError(f"unknown block type: {t!r}")
        out.append("")                                   # blank line between blocks
    return "\n".join(out)

File: static/doc_model.js
/* Editor document model <-> canonical Markdown, client side.
 *
 * A faithful port of doc_model.py in its smartquotes=False mode (the manuscript
 * editor stores raw source, so no smart-quote normalization here). Kept in lockstep
 * with the Python by test_doc_model_js.html, which round-trips the same corpus and
 * compares serializations against Python-generated expectations.
 *
 * Public API (global `DocModel`):
 *     DocModel.fromMarkdown(raw)  -> [block, ...]
 *     DocModel.toMarkdown(blocks) -> string
 *
 * Block / run schema matches doc_model.py exactly:
 *     {type:"chapter",  title:str|null, byline:str|null}
 *     {type:"part",     title:str|null}
 *     {type:"subhead",  runs:[run,...]}
 *     {type:"para",     runs:[run,...]}
 *     {type:"scene"}
 *     {type:"docblock", block_type:str, attrs:{k:v}, children:[para,...]}
 *     run = {text:str, bold:bool, italic:bool}
 */
(function (global) {
  'use strict';

  // ---- block-level patterns (mirror manuscript.py) -------------------------
  // Fenced blocks whose content is line-oriented — one source line per item /
  // per line, not wrapped into a paragraph. Mirrors manuscript.LINE_BLOCKS.
  var LINE_BLOCKS = ['list', 'center', 'centre', 'right', 'left'];
  var SCENE_BREAK_RE = /^\s*(\*\s*\*\s*\*|\*{3,}|-{3,}|#{3,})\s*$/;
  var CHAPTER_RE     = /^#\s+(.*)$/;
  var SUBHEAD_RE     = /^##\s+(.*)$/;
  var DOCBLOCK_RE    = /^\s*~~~(.*)/;
  var PART_RE        = /^\s*===\s*(.*)/;
  var ATTR_RE        = /(\w+)="([^"]*)"/g;
  var BYLINE_SEP     = ' | ';   // "# Title | Author" anthology byline (mirror manuscript.py)

  function splitByline(titleLine) {
    // "The Lottery | Shirley Jackson" -> {title, byline}; no separator -> byline null.
    if (titleLine == null) return { title: null, byline: null };
    var i = titleLine.indexOf(BYLINE_SEP);
    if (i !== -1) {
      var title = titleLine.slice(0, i).trim();
      var byline = titleLine.slice(i + BYLINE_SEP.length).trim();
      return { title: title || null, byline: byline || null };
    }
    return { title: titleLine.trim() || null, byline: null };
  }

  // ---- inline emphasis patterns (mirror doc_model.py; bold before italic) --
  var BOLD_RE         = /\*\*(.+?)\*\*/g;
  var ITALIC_STAR_RE  = /(?<!\*)\*(?!\s)(.+?)(?<!\s)\*(?!\*)/g;
  var ITALIC_UNDER_RE = /_(?!\s)(.+?)(?<!\s)_/g;

  // ---- escape layer (private-use codepoints, same scheme as Python) --------
  var BSL = '\\';
  var PARK_STAR = '', PARK_UNDER = '', PARK_BSL = '';

  function parkEscapes(text) {
    return text.split(BSL + BSL).join(PARK_BSL)   // \\ -> literal backslash
               .split(BSL + '*').join(PARK_STAR)  // \* -> literal *
               .split(BSL + '_').join(PARK_UNDER); // \_ -> literal _
  }
  function restoreEscapes(text) {
    return text.split(PARK_STAR).join('*')
               .split(PARK_UNDER).join('_')
               .split(PARK_BSL).join(BSL);
  }
  function escapeAll(text) {
    return text.split(BSL).join(BSL + BSL)
               .split('*').join(BSL + '*')
               .split('_').join(BSL + '_');
  }

  function run(text, bold, italic) {
    return { text: text, bold: !!bold, italic: !!italic };
  }

  // ---- inline: markdown <-> runs -------------------------------------------

  function eachMatch(re, str, fn) {
    re.lastIndex = 0;
    var m;
    while ((m = re.exec(str)) !== null) {
      fn(m);
      if (m.index === re.lastIndex) re.lastIndex++;  // guard zero-width
    }
  }

  function splitBySpans(str, re) {
    // returns [{hit:bool, text:str}, ...] carving out regex group-1 spans
    var out = [], pos = 0;
    eachMatch(re, str, function (m) {
      if (m.index > pos) out.push({ hit: false, text: str.slice(pos, m.index) });
      out.push({ hit: true, text: m[1] });
      pos = m.index + m[0].length;
    });
    if (pos < str.length) out.push({ hit: false, text: str.slice(pos) });
    return out;
  }

  function splitItalics(seg) {
    // *...* first, then _..._ inside still-plain pieces
    var out = [];
    splitBySpans(seg, ITALIC_STAR_RE).forEach(function (p) {
      if (p.hit) { out.push({ italic: true, text: p.text }); return; }
      splitBySpans(p.text, ITALIC_UNDER_RE).forEach(function (q) {
        out.push({ italic: q.hit, text: q.text });
      });
    });
    return out;
  }

  function coalesce(runs) {
    var out = [];
    runs.forEach(function (r) {
      var last = out[out.length - 1];
      if (last && last.bold === r.bold && last.italic === r.italic) last.text += r.text;
      else out.push({ text: r.text, bold: r.bold, italic: r.italic });
    });
    return out.filter(function (r) { return r.text !== ''; });
  }

  function parseInline(text) {
    text = parkEscapes(text);
    var runs = [];
    splitBySpans(text, BOLD_RE).forEach(function (seg) {
      if (seg.hit) { runs.push(run(seg.text, true, false)); return; }
      splitItalics(seg.text).forEach(function (it) {
        if (it.text) runs.push(run(it.text, false, it.italic));
      });
    });
    runs = coalesce(runs);
    runs.forEach(function (r) { r.text = restoreEscapes(r.text); });
    return runs;
  }

  function runsToMd(runs) {
    return runs.map(function (r) {
      if (r.bold)   return '**' + escapeAll(r.text) + '**';
      if (r.italic) return '*' + escapeAll(r.text) + '*';
      return escapePlain(r.text);
    }).join('');
  }

  function sameRuns(a, b) {
    if (a.length !== b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (a[i].text !== b[i].text || a[i].bold !== b[i].bold || a[i].italic !== b[i].italic)
        return false;
    }
    return true;
  }

  function escapePlain(text) {
    // Escape only if the literal text would otherwise re-parse as emphasis.
    var escapedBsl = text.split(BSL).join(BSL + BSL);
    if (sameRuns(parseInline(escapedBsl), [run(text, false, false)])) return escapedBsl;
    return escapedBsl.split('*').join(BSL + '*').split('_').join(BSL + '_');
  }

  function isBlockLine(line) {
    return SCENE_BREAK_RE.test(line) || CHAPTER_RE.test(line)
        || SUBHEAD_RE.test(line) || DOCBLOCK_RE.test(line) || PART_RE.test(line);
  }

  function parseBlockHeader(header) {
    var parts = header.trim().split(/\s+/);
    if (!header.trim()) return { blockType: '', attrs: {} };
    var blockType = parts[0].toLowerCase();
    var rest = header.trim().slice(parts[0].length);
    var attrs = {}, m;
    ATTR_RE.lastIndex = 0;
    while ((m = ATTR_RE.exec(rest)) !== null) attrs[m[1]] = m[2];
    return { blockType: blockType, attrs: attrs };
  }

  // ---- markdown -> document model ------------------------------------------

  function fromMarkdown(raw) {
    var lines = raw.replace(/\r\n/g, '\n').replace(/\r/g, '\n').split('\n');
    var blocks = [];
    var paraBuf = [];
    var inBlock = false, blockChildren = [], blockParaBuf = [], blockType = '', blockAttrs = {};

    function flushPara() {
      if (paraBuf.length) {
        var joined = paraBuf.map(function (s) { return s.trim(); }).join(' ').trim();
        if (joined) blocks.push({ type: 'para', runs: parseInline(joined) });
        paraBuf = [];
      }
    }
    function flushBlockPara() {
      if (blockParaBuf.length) {
        if (blockType === 'poem') {
          // Verse: each source line is a line; the group is one stanza.
          var lines = [];
          blockParaBuf.forEach(function (s) {
            if (s.trim()) lines.push({ runs: parseInline(s.trim()) });
          });
          if (lines.length) blockChildren.push({ type: 'stanza', lines: lines });
        } else if (LINE_BLOCKS.indexOf(blockType) !== -1) {
          // One line = one child (a list item, or a line of an aligned block)
          blockParaBuf.forEach(function (s) {
            if (s.trim()) blockChildren.push({ type: 'para', runs: parseInline(s.trim()) });
          });
        } else {
          var joined = blockParaBuf.map(function (s) { return s.trim(); }).join(' ').trim();
          if (joined) blockChildren.push({ type: 'para', runs: parseInline(joined) });
        }
        blockParaBuf = [];
      }
    }
    function closeBlock() {
      flushBlockPara();
      blocks.push({ type: 'docblock', block_type: blockType,
                    attrs: shallow(blockAttrs), children: blockChildren.slice() });
      blockChildren = []; blockType = ''; blockAttrs = {}; inBlock = false;
    }

    for (var li = 0; li < lines.length; li++) {
      var line = lines[li];

      if (!inBlock && line.charAt(0) === BSL && isBlockLine(line.slice(1))) {
        paraBuf.push(line.slice(1));
        continue;
      }
      if (!inBlock) {
        var mPart = PART_RE.exec(line);
        if (mPart) {
          flushPara();
          blocks.push({ type: 'part', title: mPart[1].trim() || null });
          continue;
        }
      }
      var mDoc = DOCBLOCK_RE.exec(line);
      if (mDoc) {
        if (inBlock) { closeBlock(); }
        else {
          flushPara();
          var hdr = parseBlockHeader(mDoc[1]);
          blockType = hdr.blockType; blockAttrs = hdr.attrs; inBlock = true;
        }
        continue;
      }
      if (inBlock) {
        if (line.trim() === '') flushBlockPara();
        else blockParaBuf.push(line);
        continue;
      }
      var mCh = CHAPTER_RE.exec(line), mSub = SUBHEAD_RE.exec(line);
      if (mCh) {
        flushPara();
        var sb = splitByline(mCh[1]);
        blocks.push({ type: 'chapter', title: sb.title, byline: sb.byline });
        continue;
      }
      if (SCENE_BREAK_RE.test(line)) { flushPara(); blocks.push({ type: 'scene' }); continue; }
      if (mSub) { flushPara(); blocks.push({ type: 'subhead', runs: parseInline(mSub[1].trim()) }); continue; }
      if (line.trim() === '') flushPara();
      else paraBuf.push(line);
    }
    if (inBlock) closeBlock();
    flushPara();
    return blocks;
  }

  function shallow(o) { var r = {}; for (var k in o) if (o.hasOwnProperty(k)) r[k] = o[k]; return r; }

  // ---- document model -> markdown ------------------------------------------

  function toMarkdown(blocks) {
    var out = [];
    blocks.forEach(function (b) {
      if (b.type === 'chapter') {
        var title = b.title || '';
        if (title && b.byline) out.push('# ' + title + BYLINE_SEP + b.byline);
        else out.push('# ' + title);
      } else if (b.type === 'part') {
        out.push('=== ' + (b.title || ''));
      } else if (b.type === 'subhead') {
        out.push('## ' + runsToMd(b.runs));
      } else if (b.type === 'para') {
        var line = runsToMd(b.runs);
        if (isBlockLine(line)) line = BSL + line;
        out.push(line);
      } else if (b.type === 'scene') {
        out.push('* * *');
      } else if (b.type === 'docblock') {
        var header = '~~~';
        if (b.block_type) {
          header += ' ' + b.block_type;
          for (var k in b.attrs) if (b.attrs.hasOwnProperty(k)) header += ' ' + k + '="' + b.attrs[k] + '"';
        }
        out.push(header);
        var kids = b.children || [];
        if (b.block_type === 'poem') {
          for (var si = 0; si < kids.length; si++) {     // stanzas of verse lines
            if (si > 0) out.push('');                    // blank line between stanzas
            kids[si].lines.forEach(function (ln) { out.push(runsToMd(ln.runs)); });
          }
        } else if (LINE_BLOCKS.indexOf(b.block_type) !== -1) {
          for (var li = 0; li < kids.length; li++) out.push(runsToMd(kids[li].runs));
        } else {
          for (var i = 0; i < kids.length; i++) {
            out.push(runsToMd(kids[i].runs));
            if (i < kids.length - 1) out.push('');
          }
        }
        out.push('~~~');
      } else {
        throw new Error('unknown block type: ' + b.type);
      }
      out.push('');
    });
    return out.join('\n');
  }

  global.DocModel = {
    fromMarkdown: fromMarkdown,
    toMarkdown: toMarkdown,
    // exposed for tests
    _parseInline: parseInline,
    _run: run
  };
})(typeof window !== 'undefined' ? window : this);
